File: app/src/main/java/io/github/javiewer/activity/StartActivity.java
package io.github.javiewer.activity;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.AssetManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.karumi.dexter.Dexter;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionDeniedResponse;
import com.karumi.dexter.listener.PermissionGrantedResponse;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.single.PermissionListener;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.URISyntaxException;

import io.github.javiewer.Configurations;
import io.github.javiewer.JAViewer;
import io.github.javiewer.Properties;
import io.github.javiewer.R;
import io.github.javiewer.adapter.item.DataSource;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;

/**
 * StartActivity
 *
 * @author lao
 * @date 2020/9/4
 * Profile: 开始检查更新界面
 */

public class StartActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_start);

        checkPermissions(); //检查权限，创建配置
    }

    public void readProperties() {

        //读取数据更新包的下载源
        final String url = "https://raw.githubusercontent.com/ipcjs/JAViewer/master/app/src/main/assets/properties.json";
        Request request = new Request.Builder()
                .url("http://120.24.61.225:8080/talker/properties.json?t=" + System.currentTimeMillis() / 1000)
                .build();
        JAViewer.HTTP_CLIENT.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                readLocalProperties();
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                final Properties properties = JAViewer.parseJson(Properties.class, response.body().string());
                if (properties != null) {
                    new Handler(Looper.getMainLooper()).post(new Runnable() {
                        @Override
                        public void run() {
                            handleProperties(properties);
                        }
                    });
                }
            }
        });
    }

    public void readLocalProperties() {

        try {
            AssetManager assetManager = getAssets(); //获得assets资源管理器（assets中的文件无法直接访问，可以使用AssetManager访问）
            InputStreamReader inputStreamReader = new InputStreamReader(assetManager.open("properties.json"),"UTF-8"); //使用IO流读取json文件内容
            BufferedReader br = new BufferedReader(inputStreamReader);//使用字符高效流
            String line;
            StringBuilder builder = new StringBuilder();
            while ((line = br.readLine())!=null){
                builder.append(line);
            }
            br.close();
            inputStreamReader.close();

            final Properties properties = JAViewer.parseJson(Properties.class, builder.toString());
            if (properties != null) {
                new Handler(Looper.getMainLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        handleProperties(properties);
                    }
                });
            }

/*            JSONObject testJson = new JSONObject(builder.toString()); // 从builder中读取了json中的数据。
            // 直接传入JSONObject来构造一个实例
            JSONArray array = testJson.getJSONArray("banks");

            Log.e("banks",array.toString());

            for (int i = 0;i<array.length();i++){
                JSONObject jsonObject = array.getJSONObject(i);
                String text = jsonObject.getString("text");
                String value = jsonObject.getString("value");
                Log.e("tag", "initData: "+text+value);
            }*/

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void handleProperties(Properties properties) {
        JAViewer.DATA_SOURCES.clear();
        JAViewer.DATA_SOURCES.addAll(properties.getDataSources());

        JAViewer.hostReplacements.clear();
        for (DataSource source : JAViewer.DATA_SOURCES) {
            try {
                String host = new URI(source.getLink()).getHost();
                for (String h : source.legacies) {
                    JAViewer.hostReplacements.put(h, host);
                }
            } catch (URISyntaxException e) {
                e.printStackTrace();
            }
        }

        int currentVersion;
        try {
            currentVersion = this.getPackageManager().getPackageInfo(this.getPackageName(), 0).versionCode;
        } catch (PackageManager.NameNotFoundException e) {
            throw new RuntimeException("Hacked???");
        }

        if (properties.getLatestVersionCode() > 0 && currentVersion < properties.getLatestVersionCode()) {

            String message = "新版本：" + properties.getLatestVersion();
            if (properties.getChangelog() != null) {
                message += "\n\n更新日志：\n\n" + properties.getChangelog() + "\n";
            }

            final AlertDialog dialog = new AlertDialog.Builder(this)
                    .setTitle("发现更新")
                    .setMessage(message)
                    .setNegativeButton("忽略更新", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            start();
                        }
                    })
                    .setPositiveButton("更新", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            start();
                            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/ccclao/JAViewer/releases")));
                        }
                    })
                    .create();
            dialog.show();
        } else {
            start();
        }

    }

    public void start() {
        startActivity(new Intent(StartActivity.this, MainActivity.class));
        finish();
    }

    private void checkPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && this.checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            Dexter.withActivity(this)
                    .withPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    .withListener(new PermissionListener() {
                        @Override
                        public void onPermissionGranted(PermissionGrantedResponse response) {
                            checkPermissions();
                        }

                        @Override
                        public void onPermissionDenied(PermissionDeniedResponse response) {
                            new AlertDialog.Builder(StartActivity.this)
                                    .setTitle("权限申请")
                                    .setCancelable(false)
                                    .setMessage("JAViewer 需要储存空间权限，储存用户配置。请您允许。")
                                    .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                                        @Override
                                        public void onClick(DialogInterface dialogInterface, int i) {
                                            checkPermissions();
                                        }
                                    })
                                    .show();
                        }

                        @Override
                        public void onPermissionRationaleShouldBeShown(PermissionRequest permission, PermissionToken token) {
                            token.continuePermissionRequest();
                        }
                    })
                    .onSameThread()
                    .check();
            return;
        }

        File oldConfig = new File(StartActivity.this.getExternalFilesDir(null), "configurations.json");
        File config = new File(JAViewer.getStorageDir(), "configurations.json");
        if (oldConfig.exists()) {
            oldConfig.renameTo(config);
        }

        File noMedia = new File(JAViewer.getStorageDir(), ".nomedia");
        try {
            noMedia.createNewFile();
        } catch (IOException e) {
            e.printStackTrace();
        }

        JAViewer.CONFIGURATIONS = Configurations.load(config);

        readProperties();
    }

}

File: app/src/main/java/io/github/javiewer/activity/StartActivity.kt
package io.github.javiewer.activity

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import io.github.javiewer.JAViewer
import io.github.javiewer.R
import io.github.javiewer.data.migration.JsonToRoomMigrator
import io.github.javiewer.repository.DataSourceRepository
import io.github.javiewer.repository.PropertiesRepository
import kotlinx.coroutines.launch
import java.io.File
import java.io.IOException
import javax.inject.Inject

/**
 * 启动/闪屏 Activity，负责应用初始化流程。
 *
 * 初始化步骤：
 * 1. 加载旧版 JSON 配置文件
 * 2. 触发 JSON 到 Room 的数据迁移
 * 3. 从 GitHub 获取远程配置（失败时回退到本地）
 * 4. 初始化数据源和域名重写映射
 * 5. 检查应用更新
 * 6. 跳转到 [MainActivity]
 *
 * 使用 Hilt 注入 [PropertiesRepository]、[DataSourceRepository]、[JsonToRoomMigrator]。
 */
@AndroidEntryPoint
class StartActivity : AppCompatActivity() {

    @Inject lateinit var propertiesRepository: PropertiesRepository
    @Inject lateinit var dataSourceRepository: DataSourceRepository
    @Inject lateinit var migrator: JsonToRoomMigrator

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_start)
        init()
    }

    /** 初始化应用：加载配置、触发迁移、获取远程属性 */
    private fun init() {
        val config = File(JAViewer.getStorageDir(this), "configurations.json")

        val noMedia = File(JAViewer.getStorageDir(this), ".nomedia")
        try {
            noMedia.createNewFile()
        } catch (_: IOException) {
        }

        JAViewer.CONFIGURATIONS = io.github.javiewer.Configurations.load(config)
        lifecycleScope.launch {
            migrator.migrateIfNeeded()
            loadProperties()
        }
    }

    /** 异步加载远程配置 */
    private fun loadProperties() {
        lifecycleScope.launch {
            val properties = propertiesRepository.fetchProperties()
            if (properties != null && !isFinishing) {
                handleProperties(properties)
            }
        }
    }

    /** 处理远程配置：初始化数据源、检查更新 */
    private fun handleProperties(properties: io.github.javiewer.Properties) {
        if (isFinishing || isDestroyed) return

        val sources = properties.getDataSources()
        dataSourceRepository.setDataSources(sources)
        JAViewer.DATA_SOURCES.clear()
        JAViewer.DATA_SOURCES.addAll(sources)

        JAViewer.hostReplacements.clear()
        JAViewer.hostReplacements.putAll(dataSourceRepository.getHostReplacements())

        val currentVersion = try {
            val info = packageManager.getPackageInfo(packageName, 0)
            androidx.core.content.pm.PackageInfoCompat.getLongVersionCode(info).toInt()
        } catch (_: PackageManager.NameNotFoundException) {
            0 // 无法获取版本号时返回默认值
        }

        if (properties.getLatestVersionCode() > 0 && currentVersion < properties.getLatestVersionCode()) {
            var message = "新版本：${properties.getLatestVersion()}"
            properties.changelog?.let {
                message += "\n\n更新日志：\n\n$it\n"
            }
            AlertDialog.Builder(this)
                .setTitle("发现更新")
                .setMessage(message)
                .setNegativeButton("忽略更新") { _, _ -> start() }
                .setPositiveButton("更新") { _, _ ->
                    start()
                    startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/ccclao/JAViewer/releases")))
                }
                .create()
                .show()
        } else {
            start()
        }
    }

    /** 跳转到主页 */
    private fun start() {
        JAViewer.recreateService()
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }
}
